### PR TITLE
Tracing worker

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following is the tracing worker. This worker will manage the
lifecycle of a tracked tracer. A traced tracer is a wrapper around
a client for a given namespace (model or controller namespace), so
that if a tracer client fails, it doesn't bring down every other
tracer. As we're using the runner package, a new client will be
spawned and a tracer can be selected again.

This brings in a new dependency: go.opentelemetry.io/otel. This is
the default standard for open telemetry, although it does have a
few rough edges.

 1. There is no logging support (assoicating logs with spans)
 2. There is a global instance logger, which we will have to work around

Hopefully point 1 is actively being worked on and might arrive in
the not to distant future.

As this is just setting up the worker, there isn't any manual tests
to add to this PR, that will come later. For now it should bootstrap
and react to the changes in the agent/controller config.

The implementation mirrors the dbaccessor, including the new error
catacomb.ErrDying support.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

You can now test the tracing worker is up and running by tailing the controller config.

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju deploy ubuntu
$ juju controller-config "open-telemetry-endpoint=foo;open-telemetry-enabled=true"
```

### CODE REVIEW

File: worker/dbaccessor/worker_integration_test.go [hunk 1]
```diff
@@ -69,6 +69,7 @@ var _ = gc.Suite(&integrationSuite{})
 
 func (s *integrationSuite) SetUpTest(c *gc.C) {
 	s.DqliteSuite.SetUpTest(c)
+	s.dqliteAppIntegrationSuite.SetUpTest(c)
```
Comment (manadart) [inline worker/dbaccessor/worker_integration_test.go:72]: `DqliteSuite.SetUpTest` should be called in `dqliteAppIntegrationSuite.SetUpTest` as it is embedded in that struct. Composing `SetUpTest` from multiple embedded layers is a smell.

File: worker/tracing/tracer.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	return t.stackTracesEnabled
+}
+
+func (t *tracer) loop() error {
+	defer func() {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+		defer cancel()
+
+		if err := t.client.Stop(ctx); err != nil {
+			t.logger.Infof("failed to stop client: %v", err)
```
Comment (manadart) [inline worker/tracing/tracer.go:121]: Are these worthy of `Errorf`?

File: worker/tracing/tracer.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	defer s.cancel()
+
+	s.span.SetAttributes(attributes(attrs)...)
+	s.span.End(trace.WithStackTrace(s.stackTracesEnabled))
+}
+
+func attributes(attrs []coretracing.Attribute) []attribute.KeyValue {
+	kv := make([]attribute.KeyValue, len(attrs))
+	for _, attr := range attrs {
+		kv = append(kv, attribute.String(attr.Key(), attr.Value()))
```
Comment (manadart) [inline worker/tracing/tracer.go:210]: Shame they are not using generics; we could too.

File: worker/tracing/worker.go [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+	return w.catacomb.Wait()
+}
+
+// GetTracer returns a tracer for the given namespace.
+func (w *tracerWorker) GetTracer(namespace string) (coretracing.Tracer, error) {
+	// First check if we've already got the tracer worker already running. If
+	// we have, then return out quickly. The tracerRunner is the cache, so there
+	// is no need to have a in-memory cache here.
+	if tracer, err := w.workerFromCache(namespace); err != nil {
+		if errors.Is(err, w.catacomb.ErrDying()) {
```
Comment (manadart) [inline worker/tracing/worker.go:173]: All these lessons learned...

File: worker/tracing/manifold_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package tracing
```
Comment (manadart) [inline worker/tracing/manifold_test.go:4]: I'd consider just `trace`. We have lots of {verb}er; this is {verb}ing...

**Review by manadart — APPROVED** [review]
A few comments, but this looks good.

File: worker/tracing/tracer.go (hunk 2 above)
Comment (SimonRichardson) [reply worker/tracing/tracer.go:121]: Error indicates you should perform an action when this happens, nothing we can do here.

File: worker/tracing/manifold_test.go (hunk 5 above)
Comment (SimonRichardson) [reply worker/tracing/manifold_test.go:4]: Sure.

File: worker/tracing/tracer.go (hunk 3 above)
Comment (SimonRichardson) [reply worker/tracing/tracer.go:210]: Indeed

File: worker/dbaccessor/worker_integration_test.go (hunk 1 above)
Comment (SimonRichardson) [reply worker/dbaccessor/worker_integration_test.go:72]: I'd agree, but you're actively not calling the dqliteAppIntegrationSuite parent suite. So you need to call setup test to get the skip to behave correctly. It's smell, but it's only used in this one location.

File: core/trace/tracer.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/juju/errors"
+)
+
+const (
+	// ErrTracerDying is used to indicate to *third parties* that the
+	// tracer worker is dying, instead of catacomb.ErrDying, which is
+	// unsuitable for propagating inter-worker.
+	// This error indicates to consuming workers that their dependency has
+	// become unmet and a restart by the dependency engine is imminent.
+	ErrTracerDying = errors.ConstError("tracer worker is dying")
```
Comment (tlm) [inline core/trace/tracer.go:18]: Is this the right place for this to live? Considering it pertains to the functionality of the worker.

File: core/trace/tracer.go
Comment (tlm) [inline core/trace/tracer.go:1]: This is a thing of beauty.

**Review by tlm — COMMENTED** [review]
Thanks @SimonRichardson, this looks like some awesome work.

**Pull request merged**